### DISCUSSION TITLE
Road network-based routes between stops?

### DISCUSSION THREAD

**Discussion by jimmy-feng** [anitagraser/movingpandas#208] [Q&A]
Hi,

Apologies if this has already been addressed but I couldn't find any relevant documentation online.

My question is: does the package provide functions for determining routes based on the road network? If not, could you provide suggestions for possible implementation?

For more context, I have a dataset on human travels that look like this:
![image](https://user-images.githubusercontent.com/52459230/157537832-6775a60f-b595-44c6-a232-7286fec3250c.png)

In my dataset, I have the lat, lon, time, and unique id of each stop in a person's trajectory.  My issues are that the generated trajectories from TrajectoryCollection() result in Euclidean line segments which does not align with how people get around. To work around this, I was able to extract possible routes using the [openrouteservice library](https://github.com/GIScience/openrouteservice-py) and process the data into a GeoDataFrame in which each row corresponds to a shapely.LineString from an origin to a destination (i.e. from one stop to another stop): 
![image](https://user-images.githubusercontent.com/52459230/157539273-f3f82d8f-6c87-446d-bd0b-6f819df304e8.png)

Now, I am struggling to combine my routes with my stops to produce road network-based trajectories and was wondering if anyone could provide any pointers (would MovingPandas even be suitable for my needs?) If I'm not providing enough information, please let me know!

Super thanks,
Jimmy

**Comment by anitagraser** [comment]
There is no MovingPandas example yet for this use case. However, the most likely candidate to solve this issue is the osmnx library. You can find their examples here: https://github.com/gboeing/osmnx-examples

**Reply by jimmy-feng** [reply]
Thanks for the quick response, Anita. If I'm understanding your suggestion correctly, I already have the routes but I'm wondering what would be a good way to convert them into trajectory objects -- to break up my LineString routes:
![image](https://user-images.githubusercontent.com/52459230/157543844-52795940-a846-4cd1-83af-a9b855b571f5.png)

into individual points and have TrajectoryCollection() interpolate the line segments or to insert the LineString itself and replace the interpolated components into an already constructed Trajectory object... 

Only one way to find out...

**Reply by anitagraser** [reply]
Oh, I see. So you already have the route geometry. Does the route also contain travel time info along the way that could be used for the trajectory?

**Reply by jimmy-feng** [reply]
Yes and no.
_Short answer:_ I have the route segments ordered sequentially between stops but their exact timings aren't accurate. I joined the route segments with a pd.DataFrame containing the stop locations (e.g., the first image in my original post) and then constructed a TrajectoryCollection from this joined DataFrame. It mostly works, except I noticed (A) a gap between two records (i.e., timestamps/stops) as well as (B) a generated connection between two timestamps that are quite distant from each other that I can't quite figure out how to resolve, with a rudimentary application of a Plotly 3D Line Plot:
![image](https://user-images.githubusercontent.com/52459230/158627031-8b2f728d-6082-4506-851b-d1513503a8b3.png)

_Long answer:_ The determined routes from openrouteservice are returned as a JSON, and included is a summary of the duration (seconds) along each line segment:
![image](https://user-images.githubusercontent.com/52459230/158622708-968c1769-1a3e-4a92-8310-58b0a560746d.png)

and each JSON contains a shapely LineString of the route that looks like this:
![image](https://user-images.githubusercontent.com/52459230/158623149-0c03cf52-e798-4096-84b6-4891de46d9c9.png)
![image](https://user-images.githubusercontent.com/52459230/158623428-28281fbb-f5f4-483c-9aac-2c7e9c220c37.png)

I construct individual line segments from pairs of coordinates in each route. But, there are more coordinate pairs and thus individual line segments (51 total) than is implicitly referred to in the summary of the duration along each line segment (8 total), so I incremented the timestamp of each succeeding segment by a minute -- I know this is inaccurate but I wanted to at least have the segments usable as Trajectory and TrajectoryCollection objects.

**Reply by anitagraser** [reply]
That's a very interesting problem. I'm not familiar with the Openrouteservice responses and whether there could be some common gotcha that causes the issues you're experiencing. As a step forward, can you create a Jupyter notebook with a small example and your code to reproduce?

**Reply by jimmy-feng** [reply]
[https://github.com/jimmy-feng/dissertation_spacetime](https://github.com/jimmy-feng/dissertation_spacetime)

I've uploaded [a notebook](https://github.com/jimmy-feng/dissertation_spacetime/blob/main/RouteTrajectories.ipynb), [the original data (it's synthetic, so no human ethics issue)](https://github.com/jimmy-feng/dissertation_spacetime/blob/main/Scenarios_Synthetic_Data_Example.csv), and [also a csv of the generated trajectories with openrouteservice](https://github.com/jimmy-feng/dissertation_spacetime/blob/main/sample_trajectories.csv). I tried to the best of my ability to document my code line by line, and I'd gladly appreciate any feedback you might have. I know my code isn't the prettiest nor the most efficient and I'm always looking for tips to improve!

Super thanks again, Anita!

**Reply by anitagraser** [reply]
Sorry for the prolonged silence. I'll have a look when I can.
